kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,193 @@
+                   and filter_config.pw_metrics.label_patterns or {}
+
+  for _, pair in ipairs(patterns) do
+    local label_kv, label_v = str_match(key, pair.pattern)
+    if label_kv then
+      local label_k = str_sub(label_kv, 0, str_find(label_kv, "="))
+      local label_k_start, _ = str_find(key, label_k)
+
+      first_label = (label_k_start < first_label) and label_k_start or first_label
+
+      table_insert(labels, { pair.label, label_v })
+    end
+  end
+
+  if first_label ~= #key then
+    name = str_sub(key, 0, first_label - 1)
+  end
+
+  return name, labels
+end
+
+
+local function parse_key(key)
+  -- TODO: parse wa. (WasmX metrics) and lua. (metrics defined in Lua land)
+  local header = { pw = "pw." }
+
+  local name = key
+  local labels = {}
+
+  local is_pw = #key > #header.pw and key:sub(0, #header.pw) == header.pw
+
+  if is_pw then
+    name, labels = parse_pw_key(key)
+  end
+
+  name = name:gsub("%.", "_")
+
+  return name, labels
+end
+
+
+local function serialize_labels(labels)
+  local buf = buf_new()
+
+  for _, pair in ipairs(labels) do
+    buf:put(fmt(',%s="%s"', pair[1], pair[2]))
+  end
+
+  buf:get(1)  -- discard trailing comma
+
+  return "{" .. buf:get() .. "}"
+end
+
+
+local function serialize_metric(m, buf)
+  buf:put(fmt("# HELP %s\n# TYPE %s %s", m.name, m.name, m.type))
+
+  if m.type == "histogram" then
+    local sum_lines_buf = buf_new()
+    local count_lines_buf = buf_new()
+
+    for _, pair in ipairs(m.labels) do
+      local count, sum = 0, 0
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels = (#labels > 0) and serialize_labels(labels) or ""
+
+      local blabels = (#labels > 0) and (slabels:sub(1, #slabels - 1) .. ",") or "{"
+
+      for _, bin in ipairs(labeled_m.value) do
+        local ub = (bin.ub ~= 4294967295) and bin.ub or "+Inf"
+        local ubl = fmt('le="%s"', ub)
+
+        count = count + bin.count
+
+        buf:put(fmt("\n%s%s %s", m.name, blabels .. ubl .. "}", count))
+      end
+
+      sum = sum + labeled_m.sum
+
+      sum_lines_buf:put(fmt("\n%s_sum%s %s", m.name, slabels, sum))
+      count_lines_buf:put(fmt("\n%s_count%s %s", m.name, slabels, count))
+    end
+
+    buf:put(sum_lines_buf:get())
+    buf:put(count_lines_buf:get())
+
+  else
+    for _, pair in ipairs(m.labels) do
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels = (#labels > 0) and serialize_labels(labels) or ""
+
+      buf:put(fmt("\n%s%s %s", m.name, slabels, labeled_m.value))
+    end
+  end
+
+  buf:put("\n")
+end
+
+
+_M.metric_data = function()
+  local i = 0
+  local metrics = {}
+  local parsed = {}
+  local buf = buf_new()
+
+  wasmx_shm.metrics:lock()
+
+  for key in wasmx_shm.metrics:iterate_keys() do
+    table_insert(metrics, { key, wasmx_shm.metrics:get_by_name(key, { prefix = false })})
+  end
+
+  wasmx_shm.metrics:unlock()
+
+  -- in WasmX the different labels of a metric are stored as separate metrics;
```

Comment:
Should we say "in Proxy-Wasm"?

Reply:
I guess it influenced our design but in the end other metrics than Proxy-Wasm's will be defined so that makes sense.